### PR TITLE
Add support for Raft consensus

### PR DESCRIPTION
**Pull request by Maelkum**
This PR adds support for execution requests to be executed on multiple nodes as part of a raft cluster.

TODOs:
- if an execution request comes in and it doesn't require consensus, it should be directly executed on the worker node, without the extra hassle of a raft cluster (worker node supports this, head node needs small tweaks)
- logger update - raft works with hclog, not zerolog
- support multiple raft clusters at the same time
- tests
- minor tweaks and usability improvements

These TODOs will be added in the following days in smaller PRs.

## Aggregation

Head node REST API aggregates the results it received from the nodes.

<details>
  <summary>Example of an execution with a three node cluster where one node failed:</summary>
  
```json
{
  "code": "200",
  "request_id": "ca5c680d-a6f6-4681-9e34-d220f469a1d6",
  "results": [
    {
      "result": {
        "stdout": "Hello, world!\n",
        "stderr": "",
        "exit_code": 0
      },
      "peers": [
        "12D3KooWHThfyFJp2z2HFKMvUy3aYNTgqgYbnn1FV8tEizmRQGnS",
        "12D3KooWJwzdgVWkycaTA7pZBzMRRpT5GEAeL9XTzEUXUjTRz8JS"
      ],
      "frequency": 66.66666666666667
    },
    {
      "result": {
        "stdout": "some garbled stuff here",
        "stderr": "Whoops, something went wrong",
        "exit_code": 14
      },
      "peers": [
        "12D3KooWGFRA2de2nLqtYXTSzucWRtXHcmEk2SuB5LnftRdmZKhU"
      ],
      "frequency": 33.333333333333336
    }
  ],
  "cluster": {
    "peers": [
      "12D3KooWGFRA2de2nLqtYXTSzucWRtXHcmEk2SuB5LnftRdmZKhU",
      "12D3KooWHThfyFJp2z2HFKMvUy3aYNTgqgYbnn1FV8tEizmRQGnS",
      "12D3KooWJwzdgVWkycaTA7pZBzMRRpT5GEAeL9XTzEUXUjTRz8JS"
    ]
  }
}
```

</details>

## Configuration

Introduced a `ValidateConfig()` method on the node to validate its configuration options, including raft cluster settings.

## Threshold

If not specified, we use a 60% execution response threshold. It means that for N nodes in the cluster, we expect 60% of the nodes to respond. At the moment we only expect responses, and not that those responses match. We still return the result nonetheless, we just change the `code` field value from `code.OK` to `code.PartialContent`.

## Head and Worker execution flows split

Since there are now growing differences between the head and worker node execution flow, figured it was cleaner to separate them now, than to shoehorn them into the same shape by force.

### CODE REVIEW

File: api/execute_test.go [hunk 1]
```diff
@@ -17,7 +18,33 @@ import (
...
-	srv := setupAPI(t)
+	executionResult := execute.Result{
+		Result: execute.RuntimeOutput{
+			Stdout:   "dummy-failed-execution-result",
+			Stderr:   "dummy-failed-execution-log",
+			ExitCode: 0,
+		},
+	}
+	peerIDs := []peer.ID{
+		peer.ID([]byte{0x0, 0x24, 0x8, 0x1, 0x12, 0x20, 0x56, 0x77, 0x86, 0x82, 0x76, 0xa, 0xc5, 0x9, 0x63, 0xde, 0xe4, 0x31, 0xfc, 0x44, 0x75, 0xdd, 0x5a, 0x27, 0xee, 0x6b, 0x94, 0x13, 0xed, 0xe2, 0xa3, 0x6d, 0x8a, 0x1d, 0x57, 0xb6, 0xb8, 0x91}),
```
Comment (dmikey) [inline api/execute_test.go:29]: what does this byte sequence do? can we add comments

**Review by dmikey — APPROVED** [review]

File: api/execute_test.go (hunk 1 above)
Comment (Maelkum) [reply api/execute_test.go:29]: Oh this is just some sample `peer.ID`, the same one we have in the `mocks` package. Will replace it with that one. So just a value for the variable.

File: node/config.go [hunk 2]
```diff
@@ -1,8 +1,13 @@
 package node
 
 import (
+	"errors"
+	"fmt"
+	"path/filepath"
 	"time"
 
+	"github.com/hashicorp/raft"
```
[the remark this answers is not in the document]
Comment (Maelkum) [reply node/config.go:9]: I think we'll probably be able to abstract it away, yes. Didn't want to go overboard on the first try until stuff is working

File: node/roll_call.go [hunk 3]
```diff
@@ -29,8 +29,13 @@ func (n *Node) processRollCall(ctx context.Context, from peer.ID, payload []byte
 	}
 	req.From = from
 
-	n.log.Debug().Str("cid", req.FunctionID).Str("request_id", req.RequestID).
-		Msg("received roll call request")
+	n.log.Debug().Str("cid", req.FunctionID).Str("request_id", req.RequestID).Str("origin", req.Origin.String()).Msg("received roll call request")
+
+	// TODO: (raft) temporary measure - at the moment we don't support multiple raft clusters on the same node at the same time.
+	if req.ConsensusNeeded && len(n.clusters) > 0 {
```
[the remark this answers is not in the document]
Comment (Maelkum) [reply node/roll_call.go:35]: I'll look into the multiplexer you sent, this is one of the first things on my list now. Would love to remove this limitation but didn't want any unexpected behavior in the wild so figured I check it beforehand and treat it as not supported.

**Pull request merged**